src/Controller/Api/ApiController.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Buddy\Repman\Controller\Api;
+
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+
+abstract class ApiController extends AbstractController
+{
+    protected function createApiForm(string $class): FormInterface
+    {
+        return $this->createForm(
+            $class, null, [
+                'allow_extra_fields' => true,
+                'csrf_protection' => false,
+            ]
+        );
+    }
+
+    /**
+     * @return array<string,mixed>
+     */
+    protected function parseJson(Request $request): array
+    {
+        if ($request->getContent() === '') {
+            return [];
+        }
+
+        $data = json_decode($request->getContent(), true);
+
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new HttpException(400, 'Invalid json');
+        }
+
+        return $data;
+    }
+
+    protected function renderFormErrors(FormInterface $form): JsonResponse
+    {
+        $errors = [];
+        foreach ($form as $child) {
+            /** @var FormError $error */
+            foreach ($child->getErrors(true) as $error) {
+                $errors[$child->getName()][] = $error->getMessage();
+            }
+        }
+
+        return $this->errors($errors);
+    }
+
+    /**
+     * @param callable $listFunction
+     *
+     * @return array<string,mixed>
+     */
+    protected function paginate($listFunction, int $total, int $perPage, int $page, string $baseUrl): array
```

Comment:
Is there a reason for not type hinting callable here? 🤔

Reply:
nope, thanks for your vigilance